src/main/java/org/generationcp/middleware/dao/AttributeDAO.java

```diff
@@ -318,4 +390,151 @@ public boolean isLocationUsedInAttribute(final Integer locationId) {
 			throw new MiddlewareQueryException(message, e);
 		}
 	}
+
+	private String createAttributeValuesQueryString(final AttributeValueSearchRequestDto attributeValueSearchRequestDto) {
+		final StringBuilder sql = new StringBuilder();
+		sql.append(ATTRIBUTE_VALUE_SELECT);
+		this.appendAttributeValuesFromQuery(sql);
+		this.appendAttributeValueSeachFilters(sql, attributeValueSearchRequestDto);
+		return sql.toString();
+	}
+
+	private void appendAttributeValuesFromQuery(final StringBuilder sql) {
+		sql.append("FROM atributs a ");
+		sql.append(" INNER JOIN germplsm g ON g.gid = a.gid ");
+		sql.append(" INNER JOIN cvterm cv ON a.atype = cv.cvterm_id ");
+		sql.append(" LEFT JOIN variable_overrides vpo ON vpo.cvterm_id = cv.cvterm_id AND vpo.program_uuid = :programUUID  ");
+		sql.append(" LEFT JOIN names ON names.gid = a.gid AND names.nstat = 1 ");
+		sql.append(" WHERE 1=1 ");
+	}
+
+	private void appendAttributeValueSeachFilters(final StringBuilder sql, final AttributeValueSearchRequestDto requestDTO) {
+		if (!CollectionUtils.isEmpty(requestDTO.getAttributeDbIds())) {
+			sql.append(" AND cv.cvterm_id IN (:attributeDbIds)");
+		}
+		if (!CollectionUtils.isEmpty(requestDTO.getAttributeNames())) {
+			sql.append(" AND IFNULL(vpo.alias, cv.name) IN (:attributeNames)");
+		}
+		if (!CollectionUtils.isEmpty(requestDTO.getAttributeValueDbIds())) {
+			sql.append(" AND a.aid IN (:attributeValueDbIds)");
+		}
+
+		if (!CollectionUtils.isEmpty(requestDTO.getDataTypes())) {
+			sql.append(" AND cv.cvterm_id IN (  SELECT vrsr.subject_id ");
+			sql.append(
+				"						FROM cvterm_relationship vrsr INNER JOIN cvterm s ON s.cvterm_id = vrsr.object_id AND vrsr.type_id = ");
+			sql.append(TermRelationshipId.HAS_SCALE.getId() + " ");
+			sql.append(
+				" 						INNER JOIN cvterm_relationship drsr ON drsr.subject_id = vrsr.object_id AND drsr.type_id = ");
+			sql.append(TermRelationshipId.HAS_TYPE.getId() + " AND drsr.object_id IN (:dataTypeIds) ) ");
+		}
+
+		if (!CollectionUtils.isEmpty(requestDTO.getGermplasmDbIds())) {
+			sql.append(" AND g.germplsm_uuid IN (:germplasmDbIds)");
+		}
+		if (!CollectionUtils.isEmpty(requestDTO.getExternalReferenceIDs())) {
+			sql.append(" AND EXISTS (SELECT * FROM external_reference_atributs ref ");
+			sql.append(" WHERE a.aid = ref.aid AND ref.reference_id IN (:referenceIDs)) ");
+		}
+
+		if (!CollectionUtils.isEmpty(requestDTO.getExternalReferenceSources())) {
+			sql.append(" AND EXISTS (SELECT * FROM external_reference_atributs ref ");
+			sql.append(" WHERE a.aid = ref.aid AND ref.reference_source IN (:referenceSources)) ");
+		}
+
+		// Search preferred names
+		if (!CollectionUtils.isEmpty(requestDTO.getGermplasmNames())) {
+			sql.append(" AND g.gid IN ( SELECT n.gid ");
```

Comment:
Can't this be simplified as "`AND nval in (:germplasmNames)`" since there's a left join already to preferred name record